ibis/pandas/udf.py

```diff
@@ -0,0 +1,192 @@
+    ...     return series.str.len() * 2
+    """
+    def wrapper(func):
+        # generate a new custom node
+
+        UDFNode = type(
+            func.__name__ + '_udf_node',
+            (ops.ValueOp,),
+            dict(input_type=input_type, output_type=output_type.array_type)
+        )
+
+        with pause_ordering():
+            @execute_node.register(
+                UDFNode, *signature(input_type, klass=pd.Series)
+            )
+            def execute_udf_node(op, *args, **kwargs):
+                return func(*args, **kwargs)
+
+            @execute_node.register(
+                UDFNode, *signature(input_type, klass=SeriesGroupBy)
+            )
+            def execute_udf_node_groupby(op, *args, **kwargs):
+                groupers = [
+                    grouper for grouper in (
+                        getattr(arg, 'grouper', None) for arg in args
+                    ) if grouper is not None
+                ]
+
+                assert all(groupers[0] == grouper for grouper in groupers[1:])
+                arguments = [getattr(arg, 'obj', arg) for arg in args]
+                groupings = groupers[0].groupings
+                return func(*arguments, **kwargs).groupby(groupings)
+
+        @functools.wraps(func)
+        def wrapped(*args):
+            return UDFNode(*args).to_expr()
+
+        return wrapped
+
+    return wrapper
+
+
+def udaf(input_type, output_type):
+    """Define a UDAF (user defined aggregation function) that takes a Pandas
+    Series as input returns a scalar.
+
+    Parameters
+    ----------
+    input_type : List[T]
+        A rules encoding the abstract type of each argument or one of the types
+        found in :mod:`~ibis.expr.datatypes`. These are found in
+        :mod:`~ibis.expr.rules`.
+    output_type : rule or ibis.expr.datatypes.DataType
+        The abstract return type of the function. This *cannot* be a rule that
+        encodes a :class:`~ibis.expr.types.ColumnExpr` since this API is for
+        defining *aggregation* functions.
+
+    Examples
+    --------
+    >>> import ibis
+    >>> @udaf(input_type=[dt.string], output_type=dt.int64)
+    ... def my_string_length_agg(series, **kwargs):
+    ...     return (series.str.len() * 2).sum()
+    """
+    def wrapper(func):
+        # generate a new custom node
+        UDAFNode = type(
+            func.__name__ + '_udaf_node',
+            (ops.Reduction,),
+            dict(input_type=input_type, output_type=output_type.scalar_type)
+        )
+
+        with pause_ordering():
+
+            @execute_node.register(
+                UDAFNode, *signature(input_type, klass=pd.Series)
+            )
+            def execute_udaf_node(op, *args, **kwargs):
+                return func(*args, **kwargs)
+
+            @execute_node.register(
+                UDAFNode, *signature(input_type, klass=SeriesGroupBy)
+            )
+            def execute_udaf_node_groupby(op, *args, context=None, **kwargs):
+                iters = (
```

Comment:
This function is a serious piece of kit (e.g. the lambda being passed in to `context.agg`), I'm not sure how much more verbose / less elegant this would be if written in a more readable fashion

Reply:
Yeah this is definitely something I want to do and is the main reason for the WIP. It's concise but pretty hard to read without knowledge of how pandas works.